### PR TITLE
Params

### PR DESCRIPTION
**Pull request by pana-cc**

### CODE REVIEW

**Comment by pana-cc** [comment]
@silkentrance https://github.com/pana-cc/mocha-typescript/blob/ff2ac4433a23a0c7c57b6f09fe4d5ea7c82561fb/test.ts#L125
the PackageTest here is enabled I think?

**Comment by silkentrance** [comment]
The PackageTest will now fail due to the missing augmentations. I will reopen #81 and we need to merge and then rebase.

**Comment by silkentrance** [comment]
Ups, I fucked up. Need to investigate.

**Comment by silkentrance** [comment]
Hm, it seems that with the augmentations not in place, the changes in this branch will fail due to interface contract issues. However, with the augmentations in place, the tests will no longer be executed.

**Comment by silkentrance** [comment]
The changes in this branch also break my existing tests for a different project of mine.
So I would say that these need to be refactored so that existing tests will not be broken.

```
/Volumes/workspaces/01-coldrye/03-es/05-solutions/screed/packages/parser-api/node_modules/mocha/lib/mocha.js:214
      throw new Error('invalid interface "' + name + '"');
            ^
Error: invalid interface "mocha-typescript"
    at Mocha.ui (/Volumes/workspaces/01-coldrye/03-es/05-solutions/screed/packages/parser-api/node_modules/mocha/lib/mocha.js:214:13)
    at Object.<anonymous> (/Volumes/workspaces/01-coldrye/03-es/05-solutions/screed/packages/parser-api/node_modules/mocha/bin/_mocha:516:7)
    at Module._compile (module.js:573:30)
    at Object.Module._extensions..js (module.js:584:10)
    at Module.load (module.js:507:32)
    at tryModuleLoad (module.js:470:12)
    at Function.Module._load (module.js:462:3)
    at Function.Module.runMain (module.js:609:10)
    at startup (bootstrap_node.js:158:16)
    at bootstrap_node.js:598:3
```

Again, running with the current changes to master it shows that everything works just fine.

**Comment by silkentrance** [comment]
@pana-cc I have added stderr assertions to test runners

**Comment by pana-cc** [comment]
When used as custom ui this line throws: https://github.com/pana-cc/mocha-typescript/blob/d1a93979f68f1f93c0496967ce426b7c1aacf676/index.ts#L361
The `context.describe.skip` is not yet populated. Few lines below:
Similar piece of code uses lambdas to delay the access to these objects: https://github.com/pana-cc/mocha-typescript/blob/d1a93979f68f1f93c0496967ce426b7c1aacf676/index.ts#L421

**Comment by dimastark** [comment]
Do you need my help?

**Comment by pana-cc** [comment]
The `PackageTests` were indeed ignoring the stderr, but some of the suites actually have failing tests, so I added `expected.err.txt` for these. The issue with running as custom ui was rather small, the global `describe`, `it` etc. are not yet available, but there was that "mark" property available that I am using to pick the proper function from the context:
https://github.com/pana-cc/mocha-typescript/blob/72167810702264ed12b9a69a2a383dc20dc305ff/index.ts#L207

**Comment by pana-cc** [comment]
@silkentrance @dimastark I am somewhat happy with the current state. Could you take a look again and check if it works on your side? If it is ok I am about to merge it. We will need additional documentation (few words in the readme) and maybe a dedicated "suite" test to this feature.

File: tslint.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
+{
+    "defaultSeverity": "error",
+    "extends": [
+        "tslint:recommended"
+    ],
+    "jsRules": {},
+    "rules": {
```
Comment (dimastark) [inline tslint.json]: Maybe `"member-access": [true, "no-public"]`?

**Comment by silkentrance** [comment]
@pana-cc I am working on a dedicated suite for the new feature.

What I found is that ``@params.only`` will cause all other tests in the same suite to be ignored.
We might want to reconsider ``@params.only`` in the context of a single test method and instead collect all params first and then just execute those that have been marked as only, or all.

**Comment by pana-cc** [comment]
I would trust on your judgement for these cases, feel free to address them and merge the PR when you are confident about it. I am happy with the value added by this PR already. ❤️

**Comment by silkentrance** [comment]
@pana-cc I think that this can be merged now. I will provide a new PR for the additional enhancements. Currently I am running in a `undefined` error and I am still figuring out why.
Additionally, I have made some changes as to how the tests are to be executed and also the typings files.

**Comment by silkentrance** [comment]
@pana-cc I will prepare a commit with the dedicated suite for the `@params` decorator.

**Comment by silkentrance** [comment]
@pana-cc somehow the replacement stuff became flaky and reports different results on different machines... so the actual results will vary from our expectations depending on which machine they are running. weird.

**Comment by silkentrance** [comment]
@pana-cc I do not know why, but the existing replacements for our expectations did no longer work as expected. I added some code to the cleanup function and then replaced the existing expected.txt with the then actual result, checking also whether it is sane and meets our requirements.

Now, finally, this can be merged 😄

**Comment by silkentrance** [comment]
@pana-cc refactoring the tests only a little bit for #90 will cause all tests to be classified as async tests.

**Pull request merged**